File: practice/algorithms/strings/two_characters/two_characters.py
#!/usr/bin/env python

def copy_substring(i_s, i_c1, i_c2):
    res_s = []
    for CHR in i_s:
        if CHR == i_c1 or CHR == i_c2:
            res_s.append(CHR)
    return res_s

def is_valid(i_s):
    p = []
    for CHR in i_s:
        if len(p) == 0 or p[-1] != CHR:
            p.append(CHR)
        else:
            p.pop()
    return len(i_s) == len(p)
    
n = int(raw_input().strip())
s = raw_input().strip()
c = list(set(s))

if (len(c)) == 1:
    print 0
else:
    res_len = 0
    for POS1 in xrange(len(c)):
        for POS2 in xrange(POS1 + 1, len(c)):
            t = copy_substring(s, c[POS1], c[POS2])
            if is_valid(t):
                res_len = max(res_len, len(t))
    print res_len

File: practice/algorithms/strings/two_characters/two_characters.cpp
#include <set>
#include <stack>
#include <string>
#include <iostream>

using namespace std;

string copy_substring(const string & i_s, char i_c1, char i_c2)
{
    string res = "";
    for (int pos = 0; pos < i_s.size(); ++pos)
    {
        if (i_s[pos] == i_c1 || i_s[pos] == i_c2)
        {
            res += i_s[pos];
        }
    }
    
    return res;
}

bool is_valid(const string & i_s)
{
    stack<char> s;
    for (int pos = 0; pos < i_s.size(); ++pos)
    {
        if (s.empty() || s.top() != i_s[pos])
        {
            s.push(i_s[pos]);
        }
        else
        {
            s.pop();
        }
    }
    
    return s.size() == i_s.size();
}

int main()
{
    int len;
    cin >> len;
    string s;
    cin >> s;
    
    set<char> c(s.begin(), s.end());
    if (c.size() == 1)
    {
        cout << 0 << endl;
    }
    else
    {
        int res = 0;
        for (set<char>::iterator it1 = c.begin(); it1 != c.end(); ++it1)
        {
            for (set<char>::iterator it2 = c.begin(); it2 != c.end(); ++it2)
            {
                string t = copy_substring(s, *it1, *it2);
                if (is_valid(t) && res < t.size())
                {
                    res = t.size();
                }
            }
        }
        
        cout << res << endl;
    }
    
    return 0;
}

